File: tests/examples/typedef.cpp
#include <functional>
#include <memory>

typedef enum { A, B, C } AEnum;
using int_type = unsigned long;

typedef struct {
    int a;
} TypeDefAst;

typedef struct StAst {
    int a;
} NamedTypeDefAst;

typedef union {
    AEnum a;
    TypeDefAst b;
} AUnion;

class EnumClass {
    AEnum a_enum;
};

class StClass {
    NamedTypeDefAst ast;
};

class StStClass {
    struct StAst ast;
};

union StUnionClass {
    char arr[10];
    NamedTypeDefAst ast;
};

class ConstClass {
    const TypeDefAst ast;
};

class UnionClass {
    AUnion a;
};

class UsingClass {
    int_type a;
};

template <typename T>
class UsingTypenameClass {
    using my = typename T::int_t;
    my a;
};

namespace abc {

template <class T>
class Template {
    T t;
};

template <typename CharType>
using UN = Template<CharType>;

template <typename U>
class UsingTemplateTypenameClass {
    using UN = abc::UN<U>;
    UN t;
};
};  // namespace abc

class UseUsingTemplateTypenameClass {
    abc::UsingTemplateTypenameClass<EnumClass> a;
};

File: tests/examples/typedef.py
from examples.utils import *

def case_typedef():
    data = read_json('typedef.cpp.json')
    expect_eq(data['TypeDefAst']['kind'] , 'struct')
    expect_eq(data['AEnum']['kind'] , 'enum')
    expect_eq(data['AUnion']['kind'] , 'union')
    expect_eq(data['NamedTypeDefAst']['kind'] , 'struct')
    expect_eq(data['NamedTypeDefAst']['typedef'] , 'StAst')
    expect_eq(data['StAst']['kind'] , 'struct')
    expect_eq(data['EnumClass']['fields'][0]['types'][0]['type'] , 'AEnum')
    expect_eq(data['StClass']['fields'][0]['types'][0]['type'] , 'StAst')
    expect_eq(data['ConstClass']['fields'][0]['types'][0]['type'] , 'TypeDefAst')
    expect_eq(data['UnionClass']['fields'][0]['types'][0]['type'] , 'AUnion')
    expect_eq(data['UsingClass']['fields'][0]['types'][0]['type'] , 'unsigned long')
    expect_false(data['abc::UsingTemplateTypenameClass']['fields'][0]['templates'][0]['isReduced'])
